toolkit/pallets/block-production-log/src/lib.rs

```diff
@@ -0,0 +1,112 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod benchmarking;
+#[cfg(test)]
+pub mod mock;
+#[cfg(test)]
+pub mod test;
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use sp_block_production_log::*;
+	use sp_consensus_slots::Slot;
+	use sp_runtime::BoundedVec;
+	use sp_std::vec::Vec;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		#[pallet::constant]
+		type MaxLogSize: Get<u32>;
+
+		#[pallet::constant]
+		type MaxProducerIdLength: Get<u32>;
+
+		type BlockProducerId: Member + Parameter + MaxEncodedLen;
+
+		type WeightInfo: WeightInfo;
+
+		fn current_slot() -> Slot;
+	}
+
+	#[pallet::storage]
+	pub type CurrentBlockProducerId<T: Config> = StorageValue<_, T::BlockProducerId, OptionQuery>;
+
+	#[pallet::storage]
+	pub type Log<T: Config> =
+		StorageValue<_, BoundedVec<(Slot, T::BlockProducerId), T::MaxLogSize>, ValueQuery>;
+
+	/// This storage protects against calling `append` multiple times for the same slot.
+	/// It is required because `Log` can be emptied and there is no other way to remember the last stored slot.
+	#[pallet::storage]
+	pub type LatestSlot<T: Config> = StorageValue<_, Slot, OptionQuery>;
```

Comment:
if we will assume that starting point is 0, then switching from `OptionQuery` to `ValueQuery` will allow to skip check if value exists